sys/security/mac_do/mac_do.c

```diff
@@ -1264,13 +1316,222 @@ SYSCTL_JAIL_PARAM_SYS_SUBNODE(mac, do, CTLFLAG_RW, "Jail MAC/do parameters");
+	}
+}
+
+static void
+hold_conf(struct mac_do_conf *conf)
+{
+	refcount_acquire(&conf->use_count);
+}
+
+static void
+drop_conf(struct mac_do_conf *conf)
+{
+	if (refcount_release(&conf->use_count)) {
+		if (conf->exec_paths_str != NULL)
+			free(conf->exec_paths_str, M_PATH);
+		free(conf, M_PATH);
+	}
+}
+
+static void
+remove_conf(struct prison *const pr)
+{
+	struct mac_do_conf *old_conf;
+	int error __unused;
+
+	prison_lock(pr);
+
+	old_conf = osd_jail_get(pr, conf_osd_jail_slot);
+	error = osd_jail_set(pr, conf_osd_jail_slot, NULL);
+	MPASS(error == 0);
+
+	osd_jail_del(pr, conf_osd_jail_slot);
+	prison_unlock(pr);
+
+	if (old_conf != NULL)
+		drop_conf(old_conf);
+}
+
+static void
+set_conf(struct prison *const pr, struct mac_do_conf *const conf)
+{
+	struct mac_do_conf *old_conf;
+	hold_conf(conf);
+
+	prison_lock(pr);
+	old_conf = osd_jail_get(pr, conf_osd_jail_slot);
+	osd_jail_set(pr, conf_osd_jail_slot, conf);
+	prison_unlock(pr);
+	if (old_conf != NULL) drop_conf(old_conf);
+}
+
+static void
+set_empty_conf(struct prison *const pr)
+{
+	struct mac_do_conf *const conf = alloc_conf();
+
+	set_conf(pr, conf);
+}
+
+static void
+parse_exec_paths_into_conf(struct mac_do_conf *conf, const char *pathstr)
+{
+	const char *start = pathstr;
+	const char *end;
+	int count = 0;
+
+	while (*start != '\0' && count < MAX_EXEC_PATHS) {
+		while (*start == ':' && *start != '\0') start++;
+
+		end = start;
+		while (*end != ':' && *end != '\0') end++;
+
+		size_t len = end - start;
+		if (len > 0 && len < MAX_EXEC_PATH_LEN && *start == '/') {
+			bcopy(start, conf->exec_paths[count], len);
+			conf->exec_paths[count][len] = '\0';
+			count++;
+		}
+
+		start = end;
+	}
+	conf->exec_path_count = count;
+}
+
+static int
+mac_do_sysctl_exec_paths(SYSCTL_HANDLER_ARGS)
+{
+	char *buf;
+	struct prison *td_pr = req->td->td_ucred->cr_prison;
+	struct mac_do_conf *conf;
+	int error;
+
+	conf = find_conf_locked(td_pr);
+
+	if (conf != NULL && conf->exec_paths_str != NULL) 
+		buf = strdup(conf->exec_paths_str, M_PATH);
+	else 
+		buf = strdup(exec_paths, M_PATH);
+
+	error = sysctl_handle_string(oidp, buf, EXEC_PATHS_MAXLEN, req);
+	if (error != 0 || req->newptr == NULL) {
+		free(buf, M_PATH);
+		return error;
+	}
+
+	const char *start = buf;
+	const char *end;
+	int count = 0;
+
+	while (*start != '\0') {
+		while (*start == ':' && *start != '\0') start++;
+
+		end = start;
+		while (*end != ':' && *end != '\0') end++;
+
+		size_t len = end - start;
+		if (len == 0) continue;
+
+		if (*start != '/') return EINVAL;
+
+		if (len >= MAX_EXEC_PATH_LEN) return EINVAL;
+
+		count++;
+		if (count > MAX_EXEC_PATHS) return EINVAL;
+
+		start = end;
+	}
+
+	if (conf != NULL) {
+		if (conf->exec_paths_str != NULL)
+			free(conf->exec_paths_str, M_PATH);
+		conf->exec_paths_str = strdup(buf, M_PATH);
+		parse_exec_paths_into_conf(conf, conf->exec_paths_str);
+	} else {
+		strlcpy(exec_paths, buf, EXEC_PATHS_MAXLEN);
+		mac_do_parse_exec_paths(exec_paths);
+	}
```

Comment:
FYI, this branch can never be taken (`conf` is never `NULL`, and `mac_do_parse_exec_paths()` is in fact dead code).